### PR TITLE
Added storages.oauth_client_id and _secret

### PR DESCRIPTION
**Pull request by fraber**
… and added an rspec to check error handling

- Added storages.oauth_client_id and storages.oauth_client_secret
  to migration, model and contract and specs
- Moved migrations to modules/storages/db/migrate
- Updated feature specs to test both error conditions

Todo:
- [ ] Create a reminder work package to use Rails 7 encryption of the secret field, once we managed to upgrade OP to Rails 7, which hopefully will be soon.

### CODE REVIEW

File: modules/storages/app/contracts/storages/storages/base_contract.rb [hunk 1]
```diff
@@ -57,12 +57,18 @@ class BaseContract < ::ModelContract
...
+    attribute :oauth_client_secret
+
+    # For parameter creator only the current user is allowed
     def validate_creator_is_user
       unless creator == user
         errors.add(:creator, :invalid)
       end
     end
 
+    # Check that the new host is accessible and running the latest version of Nextcloud.
```
Comment (wielinde) [inline modules/storages/app/contracts/storages/storages/base_contract.rb]:
```suggestion
    # Check that the new host is accessible and running a sufficiently up to date version of Nextcloud.
```

File: modules/storages/app/views/storages/admin/storages/_form.html.erb [hunk 2]
```diff
@@ -30,6 +30,7 @@ See COPYRIGHT and LICENSE files for more details.
 <!-- Included form for create and edit operations, please see the controller for comments -->
 <%= error_messages_for_contract @object, @errors %>
 
+<%= content_tag :h3, t('label_general') %>
```
Comment (wielinde) [inline modules/storages/app/views/storages/admin/storages/_form.html.erb]: AFAIK we use `fieldset` to group fields in a form. Let's stick to that.
```suggestion
 <section class="form--section">
    <fieldset class="form--fieldset">
      <legend class="form--fieldset-legend"><%= I18n.t(:label_general) %></legend>
       ...
```

File: modules/storages/app/views/storages/admin/storages/_form.html.erb [hunk 3]
```diff
@@ -58,5 +59,20 @@ See COPYRIGHT and LICENSE files for more details.
       <%= t('storages.instructions.host') %>
     </span>
   </div>
+</section>
 
+<%= content_tag :h3, t(:'storages.label_oauth') %>
```
Comment (wielinde) [inline modules/storages/app/views/storages/admin/storages/_form.html.erb]: Same here. Please use `fieldset`.

File: modules/storages/app/views/storages/admin/storages/show.html.erb [hunk 4]
```diff
@@ -50,6 +50,7 @@ See COPYRIGHT and LICENSE files for more details.
   </li>
 <% end %>
 
+<%= content_tag :h3, I18n.t('label_general') %>
```
Comment (wielinde) [inline modules/storages/app/views/storages/admin/storages/show.html.erb]: `fieldset`

File: modules/storages/app/views/storages/admin/storages/show.html.erb [hunk 5]
```diff
@@ -84,3 +85,21 @@ See COPYRIGHT and LICENSE files for more details.
     </div>
   </div>
 </div>
+
+<%= content_tag :h3, I18n.t(:'storages.label_oauth') %>
```
Comment (wielinde) [inline modules/storages/app/views/storages/admin/storages/show.html.erb]: `fieldset`

File: modules/storages/app/views/storages/admin/storages/show.html.erb [hunk 6]
```diff
@@ -84,3 +85,21 @@ See COPYRIGHT and LICENSE files for more details.
...
+    <div class="attributes-key-value--key"><%= t(:'storages.label_oauth_client_id') %></div>
+    <div class="attributes-key-value--value-container">
+      <div class="attributes-key-value--value -text">
+        <span><%= @object.oauth_client_id %></span>
+      </div>
+    </div>
+    <div class="attributes-key-value--key"><%= t(:'storages.label_oauth_client_secret') %></div>
+    <div class="attributes-key-value--value-container">
+      <div class="attributes-key-value--value -text">
+        <span><%= @object.oauth_client_secret %></span>
```
Comment (wielinde) [inline modules/storages/app/views/storages/admin/storages/show.html.erb:110]: Please do not expose the secret after it was saved once.

File: modules/storages/app/views/storages/admin/storages/show.html.erb (hunk 4 above)
Comment (fraber) [reply modules/storages/app/views/storages/admin/storages/show.html.erb]: Hi, just FYI, fieldset apparently doesn't work with attributes-key-value in a show.html.erb. Instead, I've used attributes-group--header as in ./app/views/oauth/applications/show.html.erb.

File: modules/storages/config/locales/en.yml [hunk 7]
```diff
@@ -39,6 +39,8 @@ en:
       setting_up_storages: "For setting up file storages, please visit"
       setting_up_storages_non_admin: "Administrators can set up file storages in Administration / File Storages."
       all_available_storages_already_added: "All available storages are already added to the project."
+      oauth_client_id: "OAuth Client ID from attached storage (Nextcloud) that allows OpenProject to connect to the storage. This should be a lengthy string of garbled characters. Please see the storage manufacturers instructions on how to setup an OAuth application."
```
Comment (wielinde) [inline modules/storages/config/locales/en.yml]: The translation string here is Nextcloud specific but the field itself is not necessarily Nextcloud specific.
```suggestion
      oauth_client_id: "OAuth Client ID for the attached storage, i.e. Nextcloud, that allows to identify OpenProject when connecting to the storage. This should be a lengthy string of garbled characters. For obtaining the OAuth Client ID, please see the storage manufacturers instructions on how to setup an OAuth application."
```

File: modules/storages/config/locales/en.yml [hunk 8]
```diff
@@ -39,6 +39,8 @@ en:
...
+      oauth_client_secret: "OAuth Client Secret from the attached storage (Nextcloud) above."
```
Comment (wielinde) [inline modules/storages/config/locales/en.yml]: I am not sure if that instruction here adds any value to the user. How does it help the user to know what to put in that field? How does it reduce doubt?

I don't have a better phrase. Shall we simple remove it?

Comment (fraber) [reply modules/storages/config/locales/en.yml]: Meine Idee wäre es, die Nachricht durch einen einen Link auf ein Tutorial zu ergänzen, in dem für einen Admin aufgezeigt wird, wie das aufzusetzen ist. Das macht aber erst Sinn, wenn File-Links (einigermaßen) fertig ist. Vielleicht dann einen Field-Test mit 'nem unbedarften Admin?

File: modules/storages/app/controllers/storages/admin/storages_controller.rb [hunk 9]
```diff
@@ -52,7 +52,12 @@ def index
 
   # Show page with details of one Storage object.
   # Called by: Global app/config/routes.rb to serve Web page
-  def show; end
+  def show
+    # Replace the client secret by **** for security reasons
+    if @object.oauth_client_secret.present?
+      @object.oauth_client_secret = "****"
```
Comment (wielinde) [inline modules/storages/app/controllers/storages/admin/storages_controller.rb:58]: @fraber I am afraid that overwriting a model's attribute for the purpose of presentation in the view is not the best idea. Who knows that model is going to be used for next (in whatever hook) and then there is wrong data stored in it. So I'd prefer decoupling the presentation from the data here.

File: modules/storages/app/controllers/storages/admin/storages_controller.rb [hunk 10]
```diff
@@ -96,7 +101,14 @@ def create
   # Edit page is very similar to new page, except that we don't need to set
   # default attribute values because the object already exists
   # Called by: Global app/config/routes.rb to serve Web page
-  def edit; end
+  def edit
+    # Replace the client secret by **** for security reasons.
+    # A minimum length validation makes sure the old secret isn't
+    # easily overwritten.
+    if @object.oauth_client_secret.present?
+      @object.oauth_client_secret = "****"
```
Comment (wielinde) [inline modules/storages/app/controllers/storages/admin/storages_controller.rb:109]: @fraber If you do that, then the next time you save the storage you overwrite the oauth_client_secret with `****`.

File: modules/storages/spec/features/admin_storages_spec.rb [hunk 11]
```diff
@@ -48,42 +48,77 @@
...
     expect(page.find('.title-container')).to have_text('Edit: NC 1')
 
-    mock_server_capabilities_response("https://other.example.com")
+    # Test the behavior of a failed validation with code 400 (Bad Request)
+    # simulating server not running Nextcloud
+    mock_server_capabilities_response("https://other.example.com", response_code: '400')
     page.find('#storages_storage_name').set("Other NC")
     page.find('#storages_storage_host').set("https://other.example.com")
+    page.find('#storages_storage_oauth_client_id').set("23456789ABCDEF01")
+    page.find('#storages_storage_oauth_client_secret').set("3456789ABCDEF012")
```
Comment (wielinde) [inline modules/storages/spec/features/admin_storages_spec.rb:79]: @fraber Here, before the `set` the secret is probably `****`. So if you saved that form without changing the secret, the real secret would be lost.

Comment (fraber) [reply modules/storages/spec/features/admin_storages_spec.rb:79]: Hi, I've set a minimum length for client_id and secret to 10.
That way I make sure it's not overwritten by "****".

File: modules/storages/app/controllers/storages/admin/storages_controller.rb (hunk 9 above)
Comment (fraber) [reply modules/storages/app/controllers/storages/admin/storages_controller.rb:58]: Hi, this is all just temporary until Rails 7 introduces this new way of storing secrets.
I've added a check for length > 10 in the model to make sure the secret is not overwritten.

Comment (fraber) [reply modules/storages/app/controllers/storages/admin/storages_controller.rb:58]: Ok, I'll modify this and introduce a parameter in the view or something similar.

**Pull request closed without merging**